lualib/rspamadm/clickhouse.lua

```diff
@@ -0,0 +1,247 @@
+parser:option '-u --user'
+      :description 'Username to use for Clickhouse'
+      :argname('user')
+parser:option '--use-gzip'
+      :description 'Use Gzip with Clickhouse'
+      :argname('use_gzip')
+      :default(true)
+parser:flag '--use-https'
+      :description 'Use HTTPS with Clickhouse'
+      :argname('use_https')
+
+local neural_profile = parser:command 'neural_profile'
+      :description 'Generate symbols profile using data from Clickhouse'
+neural_profile:option '-l --limit'
+      :description 'Maximum number of rows to fetch from Clickhouse'
+      :argname('limit')
+      :default(100000)
+
+local http_params = {
+  config = rspamd_config,
+  ev_base = rspamadm_ev_base,
+  session = rspamadm_session,
+  resolver = rspamadm_dns_resolver,
+}
+
+local function load_config(config_file)
+  local _r,err = rspamd_config:load_ucl(config_file)
+
+  if not _r then
+    rspamd_logger.errx('cannot parse %s: %s', config_file, err)
+    os.exit(1)
+  end
+end
+
+local function handle_neural_profile(args)
+  local query = string.format(
+      "SELECT Action, Symbols.Names FROM rspamd WHERE SettingsId = '' ORDER BY TS DESC LIMIT %s",
+      args.limit)
+  local upstream = args.upstream:get_upstream_round_robin()
+  local err, res = lua_clickhouse.select_sync(upstream, args, http_params, query)
+  if err ~= nil then
+    print(string.format('Error querying Clickhouse: %s', err))
+    os.exit(1)
+  end
+
+  -- Walk results once to collect all symbols & count ocurrences
+  local known_symbols = {}
+  local symbols_count, removed, seen_total = 1, 0, 0
+  for _, r in ipairs(res) do
+    local is_spam = true
+    if r['Action'] == 'no action' or r['Action'] == 'greylist' then
+      is_spam = false
+    end
+    seen_total = seen_total + 1
+    for _, sym in ipairs(r['Symbols.Names']) do
+      local t = known_symbols[sym]
+      if not t then
+        local spam_count, ham_count = 0, 0
+        if is_spam then
+          spam_count = spam_count + 1
+        else
+          ham_count = ham_count + 1
+        end
+        known_symbols[sym] = {
+          id = symbols_count,
+	  seen = 1,
+	  seen_ham = ham_count,
+	  seen_spam = spam_count,
+        }
+	symbols_count = symbols_count + 1
+      else
+        known_symbols[sym].seen = known_symbols[sym].seen + 1
+	if is_spam then
+          known_symbols[sym].seen_spam = known_symbols[sym].seen_spam + 1
+        else
+          known_symbols[sym].seen_ham = known_symbols[sym].seen_ham + 1
+        end
+      end
+    end
+  end
+
+  local known_symbols_list = {}
+  for k, v in pairs(known_symbols) do
+    local lower_count, higher_count
+    if v.seen_spam > v.seen_ham then
+      lower_count = v.seen_ham
+      higher_count = v.seen_spam
+    else
+      lower_count = v.seen_spam
+      higher_count = v.seen_ham
+    end
+    if lower_count / higher_count >= 0.95 then
+      -- Weak ham/spam correlation, remove symbol
+      print(string.format("Removing %s - weak ham/spam correlation", k))
```

Comment:
I would suggest to use rspamd_logger.messagex here.

Reply:
It doesn't seem useful in context of rspamadm. The messages are mainly for debugging purposes so I suppose I'll get rid of many of them.